Cython/Compiler/Dataclass.py

```diff
@@ -568,32 +587,33 @@ def generate_cmp_code(code, op, funcname, node, fields):
     names = [name for name, field in fields.items() if (field.compare.value and not field.is_initvar)]
 
     with code.indenter(f"def {funcname}(self, other):"):
-        code.add_code_chunk(f"""
-            if other.__class__ is not self.__class__: return NotImplemented
+        code.add_code_line(f"cdef {node.class_name} other_cast")
+        with code.indenter(f"with {code.critical_section_placeholder_name}(self, other):"):
+            code.add_code_chunk(f"""
+                if other.__class__ is not self.__class__: return NotImplemented
```

Comment:
Why does this need a critical section? Is the access to `self.__class__` not otherwise guarded in any way?

Reply:
No - it should be fine outside the critical section. `self.__class__` is constant for our extension types.